File: 108-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if len(nums) == 0:
            return None

        midIndex = len(nums)//2
        leftNums = nums[:midIndex]
        rightNums = nums[(midIndex + 1):]

        newNode = TreeNode(nums[midIndex])
        newNode.left = self.sortedArrayToBST(leftNums)
        newNode.right = self.sortedArrayToBST(rightNums)

        return newNode

File: 108-sorted-array-to-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
  // if length of array is only 1, then just add that node
  if (nums.length === 0) {
    return null;
  }
  // get the length of the array
  // ceiling divide by 2 to split the array into 2
  var midIndex = Math.floor(nums.length/2);
  var leftNums = nums.slice(0, midIndex);
  var rightNums = nums.slice(midIndex+1);
  // ==> that index is the parent node
  var newNode = new TreeNode(nums[midIndex]);
  // left and right half:
    // recursively do the same thing
  newNode.left = sortedArrayToBST(leftNums);
  newNode.right = sortedArrayToBST(rightNums);
  
  return newNode;
};